### ISSUE TITLE
[Feature] Pluggable hidden-states transport for online speculative-training callback (colocate / NCCL-P2P)

### ISSUE THREAD

**Issue by zhubohao911** [sgl-project/sglang#25342]
### Motivation

SGLang's online speculative-training callback writes hidden states (`hidden_states`, `aux_hidden_states`, `last_hidden_states`, `target_logits`) to a Mooncake KV store keyed by UUID; the trainer process reads them back through the same store. This is the right shape for **disaggregated** layouts (trainer and engine on different machines), and a related path for **offline** capture is being added in #13868.

For **colocated** layouts — trainer and engine sharing the same GPUs (MPS, single-node multi-role) — the Mooncake hop is wasted bandwidth and serialization cost: each hidden-states tensor traverses engine GPU → Mooncake store → trainer GPU on the same device, while the two processes could exchange them directly over `dist.batch_isend_irecv` on a shared default process group.

Frameworks doing colocate RL / online speculative-training currently maintain out-of-tree patches to bypass the Mooncake writer. This request is for a small, pluggable transport surface in the spec-training callback path so that work can move upstream without sglang taking a dependency on any particular framework.

### Proposed shape (three small patch surfaces)

1. **Distributed init — join an external default PG if one already exists.**
   When the scheduler subprocess sees `dist.is_initialized()` (or a neutral sentinel env var the framework sets), skip its own `init_process_group` and derive the TP group as a contiguous slice of the existing default. This is the pattern verl, slime, and others hand-roll today.

2. **Spec-training callback writer — abstract the transport.**
   Today the callback path is hard-wired to `EagleMooncakeStore`. Abstract it behind a small interface (`send(tensors, request_id)` / `close()`), keep `MooncakeTransport` as the default, and let frameworks register their own transport (registry shape similar to the `--model-config-parser` registry added in #25050).

3. **Per-TP-rank participation.**
   The callback runs on TP rank 0 only today. For colocate P2P, every TP rank needs to participate so each engine rank can send its local shard to its paired trainer rank without a TP-0 all-gather + scatter. The default `MooncakeTransport` would continue to early-return on `rank != 0`; a transport opting into per-rank participation declares it.

### Related

- #24119 *[Discussion] Tightening SGLang's RL inner loop API* — §2 raises the colocate question for the **weight-transfer-in** direction (`update_weights_from_ipc` + `cudaIpcMemHandle`). This issue is the symmetric **hidden-states-out** ask.
- #13868 *[Feature] Add aux and last hidden state dumping for EAGLE* — same hidden states, **offline** transport (disk). With a pluggable transport surface, that PR's `HiddenStateDumper` can register as a `DiskTransport` cleanly.
- #25050 *Add `--model-config-parser` registry for pluggable config formats* — precedent for a small registry indirection.

### Related resources

Draft PR with the RFC text: #25343. A working out-of-tree implementation against v0.5.8.post1 and v0.5.10.post1 exists and exercises all three surfaces on a 4×H100 box (1 engine × TP=4 + 4 trainers × FSDP=4, GPUs shared via MPS). Happy to drive the implementation PR once the API shape is agreed.

**Comment by zhubohao911** [comment]
Tracked in https://github.com/sgl-project/sglang/pull/25343

**Comment by gaineyllc** [comment]
## Cross-node hidden-state visibility requires CPU-staged `batch_put_from`

We are running disaggregated EAGLE-3 training (SGLang on GB10 Blackwell, Aurora trainer on a separate Dell node with RTX 3090 Ti, connected via Tailscale/TCP through Mooncake).

The existing SGLang-side spec-training patch uses `put_from(key, tensor.data_ptr(), size)` with GPU device pointers. This works for **colocated** setups but **fails silently for cross-node TCP transfers** — the Mooncake transfer engine cannot DMA from a remote GPU pointer over TCP. The data is stored locally but invisible to the remote trainer node.

### Fix (disaggregated path)

When `MOONCAKE_PROTOCOL=tcp` (indicating cross-node transfer), hidden states must be staged through CPU memory before `batch_put_from`:

```python
# Cross-node: CPU-staged for TCP transfer visibility
if self._mooncake_remote_transfer:
    _hs_cpu = hidden_states.contiguous().cpu()
    keys = [f"{base_key}_hs"]
    ptrs = [_hs_cpu.data_ptr()]
    sizes = [_hs_cpu.numel() * _hs_cpu.element_size()]
    # ... same for _ids, _lhs ...
    self.eagle_mooncake_store.batch_put_from(keys, ptrs, sizes)
else:
    # Local: GPU pointers work fine, zero-copy
    self.eagle_mooncake_store.put_from(...)
```

Note: Aurora upstream (`togethercomputer/aurora`) already handles this correctly via `HostBuffer` (pinned CPU memory + `_stage_tensors_into_buffer`). The bug is specifically in the SGLang-side scheduler patch that bypasses Aurora’s buffer layer for the initial `put_from` calls.

### Suggestion for the pluggable transport RFC

The `MooncakeTransport` default implementation should:
1. Detect protocol (`tcp` vs `shm`) at init time
2. For TCP: always stage through pinned host buffers (as Aurora’s `HostBuffer` does)
3. For SHM (colocated): allow direct GPU pointer `put_from` (zero-copy)

This ensures the disaggregated path works out-of-the-box without users hitting the silent cross-node visibility failure we encountered.

### Environment
- SGLang v0.5.8.post1 with spec-training patch
- Mooncake built with `USE_REDIS=ON`, `STORE_USE_REDIS=ON`
- Metadata: `redis://` (persistent segment registration)
- Transfer: TCP over Tailscale mesh
- Segment: 32GB (4GB default causes eviction failures under sustained training load)

**Comment by gaineyllc** [comment]
## Follow-up: Benchmarks from working cross-node EAGLE-3 training with pinned-buffer adapter

We now have stable cross-node training running (27+ steps, loss 12.1 → 5.8, accuracy 0% → 11.4%) using a thin adapter layer between SGLang and Aurora’s `HostBufferPool` + `AsyncPutManager`. Here are the fetch latency comparisons:

### Transfer method benchmarks (trainer-side `fetch` time per step)

| Method | Fetch latency | Result |
|--------|-------------|--------|
| `put_from(gpu_ptr)` (original) | ∞ (hangs) | **BROKEN** — GPU device pointers invisible to remote TCP transfer |
| `batch_put_from(.cpu().data_ptr())` (naive fix) | 5–28s | Works but allocates unpinned memory every sample, blocks inference |
| Aurora `HostBufferPool` + `AsyncPutManager` | **0.7–1.5s** | Pinned DMA, async, non-blocking, zero-alloc steady-state |

The adapter pattern that works:

```python
class MooncakeEagleAdapter:
    def __init__(self, store, buffer_size=4*1024**3, pool_size=2):
        self._host_buffer_pool = HostBufferPool(buffer_size, pool_size)
        self._async_put_manager = AsyncPutManager(store, max_workers=1)
        self._copy_stream = torch.cuda.Stream()

    def put(self, key, hidden_states, input_ids, last_hidden_states=None):
        buf = self._host_buffer_pool.get_buffer()
        self._async_put_manager.wait_for_buffer(buf.ptr)
        # Stage GPU → pinned CPU on dedicated stream
        with torch.cuda.stream(self._copy_stream):
            buffer_ptrs, sizes = self._stage_tensors(buf, tensors)
            copy_done = torch.cuda.Event()
            copy_done.record()
        # Async batch_put_from on background thread
        self._async_put_manager.submit(keys, buffer_ptrs, sizes, buf.ptr, wait_event=copy_done)
```

### Additional findings for the RFC

1. **Segment size matters**: The default 4GB `MOONCAKE_GLOBAL_SEGMENT_SIZE` causes `BatchPut failed due to insufficient space` under sustained training load. 32GB works for continuous EAGLE-3 training.

2. **Redis metadata persistence eliminates startup ordering**: With `redis://` metadata, segment registrations survive master/trainer restarts. No more chicken-and-egg between SGLang and trainer.

3. **The callback failure threshold (50) is too aggressive**: If the trainer starts after SGLang, 50 failures accumulate in ~30s and spec training is permanently disabled. Suggest exponential backoff with recovery.

### Recommendation for MooncakeTransport interface

The pluggable transport should expose:
- `transport.put(key, tensors)` — handles staging + async transfer internally
- `transport.protocol` — `tcp` (remote, pinned buffers) vs `shm` (local, GPU direct)
- Constructor takes the already-connected `MooncakeDistributedStore` instance

This keeps the SGLang scheduler simple (one `transport.put()` call) while the transport implementation handles protocol-specific optimizations internally.
